modules/date-picker/main/index.coffee

```diff
@@ -92,39 +92,67 @@ isToday = (year, month, day) ->
   date.toString() is today.toString()
 
 
+toggleInputValidity = (input, dateValidityCallback, valid, type) ->
+  if input?
+    if dateValidityCallback
+      dateValidityCallback(valid, type)
+    else
+      input.classed('hx-date-error', !valid)
 
 # Checks that the start and end dates fall within the valid range and that the
 # end date is always after the start date.
 validateDates = (datepicker) ->
   _ = datepicker._
   isRangePicker = datepicker.options.selectRange
+  if datepicker.options.v2Features
+    validityFn = _.inputOnlyMode and datepicker.options.dateValidityCallback
+    toggleInputValidity(_.input, validityFn, true)
+
+    _.inputStart?.classed('hx-date-error', false)
+    _.inputEnd?.classed('hx-date-error', false)
+    if _.validRange?
+      if _.validRange.start?
+        if _.startDate < _.validRange.start
+          toggleInputValidity(_.input, validityFn, false, 'DATE_OUTSIDE_RANGE_START')
+          _.inputStart?.classed('hx-date-error', true)
+
+        if isRangePicker and _.endDate < _.validRange.start
+          _.inputEnd?.classed('hx-date-error', true)
+
+      if _.validRange.end?
+        if _.startDate > _.validRange.end
+          toggleInputValidity(_.input, validityFn, false, 'DATE_OUTSIDE_RANGE_END')
+          _.inputStart?.classed('hx-date-error', true)
+
+        if isRangePicker and _.endDate > _.validRange.end
+          _.inputEnd?.classed('hx-date-error', true)
+  else
+    if _.validRange?
+      if _.validRange.start?
+        if _.startDate < _.validRange.start
+          _.startDate = new Date(_.validRange.start.getTime())
 
-  if _.validRange?
-    if _.validRange.start?
-      if _.startDate < _.validRange.start
-        _.startDate = new Date(_.validRange.start.getTime())
-
-      if isRangePicker and _.endDate < _.validRange.start
-        _.endDate = new Date(_.validRange.start.getTime())
+        if isRangePicker and _.endDate < _.validRange.start
+          _.endDate = new Date(_.validRange.start.getTime())
 
-    if _.validRange.end?
-      if _.startDate > _.validRange.end
-        _.startDate = new Date(_.validRange.end.getTime())
+      if _.validRange.end?
+        if _.startDate > _.validRange.end
+          _.startDate = new Date(_.validRange.end.getTime())
 
-      if isRangePicker and _.endDate > _.validRange.end
-        _.endDate = new Date(_.validRange.end.getTime())
-
-  if not isRangePicker
-    _.endDate = _.startDate
-  else
-    if _.endDate < _.startDate
-      tDate = _.endDate
+        if isRangePicker and _.endDate > _.validRange.end
+          _.endDate = new Date(_.validRange.end.getTime())
 
+    if not isRangePicker
       _.endDate = _.startDate
-      _.startDate = tDate
+    else
+      if _.endDate < _.startDate
+        tDate = _.endDate
+
+        _.endDate = _.startDate
+        _.startDate = tDate
 
-  _.endDate.setHours(0, 0, 0, 0)
-  _.startDate.setHours(0, 0, 0, 0)
+    _.endDate.setHours(0, 0, 0, 0)
+    _.startDate.setHours(0, 0, 0, 0)
```

Comment:
We no longer modify the date values in this function for v2 and instead just show the date as invalid